### PR TITLE
Allow Postman contract-testing support using ensemble design

### PR DESCRIPTION
**Pull request by JulienBreux**
### Description

This pull-request is designed to iso current developments for JS and Java.

- [x] Add ensemble sub-package
- [x] Add ensemble sub-package tests
- [x] Refactor test/helper
- [x] Update README
- [x] Add ensemble Microcks image customization
- [x] Add ensemble Postman image customization
- [x] Finalize ensemble test with good and bad implem

### Related issue(s)

- https://github.com/microcks/microcks-testcontainers-go/issues/7

### CODE REVIEW

File: ensemble/ensemble.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// Add adds an option to the list
+func (co *ContainerOptions) Add(opt testcontainers.ContainerCustomizer) {
+	co.list = append(co.list, opt)
+}
+
+// MicrocksEnsembleContainers represents the ensemble of containers
+type MicrocksEnsembleContainers struct {
+	ctx context.Context
+
+	network                    *testcontainers.DockerNetwork
```
Comment (JulienBreux) [inline ensemble/ensemble.go:33]:
```suggestion
```

File: ensemble/ensemble.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	primaryArtifactFilePath    string
+	secondaryArtifactsFilePath string
```
Comment (JulienBreux) [inline ensemble/ensemble.go:35]:
```suggestion
```

TODO:  to remove

File: ensemble/ensemble.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// RunEnsembleContainers creates instances of the Microcks and necessaries tools.
+// Using sequential start to avoid resource contention on CI systems with weaker hardware.
+func RunEnsembleContainers(ctx context.Context, opts ...Option) (*MicrocksEnsembleContainers, error) {
+	var err error
+
+	ensemble := &MicrocksEnsembleContainers{
+		ctx: ctx,
+	}
```
Comment (JulienBreux) [inline ensemble/ensemble.go:61]:
```suggestion
	ensemble := &MicrocksEnsembleContainers{ctx: ctx}
```

File: ensemble/ensemble.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+	// Options
+	defaults := []Option{WithDefaultNetwork()}
+	options := append(defaults, opts...)
```
Comment (JulienBreux) [inline ensemble/ensemble.go:65]: TODO: Simplify

File: ensemble/ensemble.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	for _, opt := range options {
+		if err = opt(ensemble); err != nil {
+			return nil, err
+		}
+	}
+
+	// Microcks container
+	ensemble.microcksContainerOptions.Add(microcks.WithEnv("POSTMAN_RUNNER_URL", "http://postman:3000"))
+	ensemble.microcksContainerOptions.Add(microcks.WithEnv("TEST_CALLBACK_URL", "http://microcks:8080"))
+	ensemble.microcksContainer, err = microcks.RunContainer(ctx, ensemble.microcksContainerOptions.list...)
```
Comment (JulienBreux) [inline ensemble/ensemble.go:75]: TODO: Export as variable + const (eg: microcks.Name = "microcks", etc...) be careful with dyn port

File: ensemble/ensemble_test.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package ensemble_test
```
Comment (JulienBreux) [inline ensemble/ensemble_test.go:1]: TODO: fulfil tests :p

**Comment by lbroudoux** [comment]
Just a suggestion: shouldn't we keep `ensemble` as a top-level concept just close to `microns.go` (and maybe rename it as `microns-ensemble.go` ?) and then move `postman` and `async-minion` containers in their own sub-packages ?

**Comment by JulienBreux** [comment]
Ok, no pb, I'll move to ensemble.go. And change some signatures.

**Comment by JulienBreux** [comment]
> Just a suggestion: shouldn't we keep `ensemble` as a top-level concept just close to `microns.go` (and maybe rename it as `microns-ensemble.go` ?) and then move `postman` and `async-minion` containers in their own sub-packages ?

The main problem I see is options in the same package area (eg WithNetwork ... etc)
What's real problem to use an import like `import "microcks.io/testcontainers-go/ensemble` vs `import "microcks.io/testcontainers-go"` in your opignon?

**Comment by lbroudoux** [comment]
No, no issue here. I was unaware of this constraint in Go and was much more thinking about having a similar approach between the different languages where possible. But I agree that we should also stick to language constraints, idioms and ways of doing things. I trust you for the Go Lang part 💪

File: README.md [hunk 7]
```diff
@@ -117,4 +117,77 @@ require.False(t, testResult.Success)
...
+You can execute a `POSTMAN` test using an ensemble that way:
+
+```go
+// Build a new TestRequest.
+testRequest := client.TestRequest{
+    ServiceId:    "API Pastries:0.0.1",
+    RunnerType:   client.TestRunnerTypeOPENAPISCHEMA,
+    TestEndpoint: "http://good-impl:3001",
+    Timeout:      2000,
+}
```
Comment (JulienBreux) [inline README.md:188]: To be reviewed!

File: ensemble/ensemble.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+		if err = opt(ensemble); err != nil {
+			return nil, err
+		}
+	}
+
+	// Microcks container
+	if ensemble.postmanEnabled {
+		ensemble.microcksContainerOptions.Add(microcks.WithEnv("POSTMAN_RUNNER_URL", "http://postman:3000"))
+	}
+	ensemble.microcksContainerOptions.Add(microcks.WithEnv("TEST_CALLBACK_URL", "http://microcks:8080"))
```
Comment (lbroudoux) [inline ensemble/ensemble.go:96]: Shouldn't you use the `defaultNetworkAliasMicrocks` const of the one picked by user? (if you let this possible...)

File: ensemble/ensemble.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+	options := append(defaults, opts...)
+	for _, opt := range options {
+		if err = opt(ensemble); err != nil {
+			return nil, err
+		}
+	}
+
+	// Microcks container
+	if ensemble.postmanEnabled {
+		ensemble.microcksContainerOptions.Add(microcks.WithEnv("POSTMAN_RUNNER_URL", "http://postman:3000"))
```
Comment (lbroudoux) [inline ensemble/ensemble.go:94]: Shouldn't you use the `defaultNetworkAliasPostman` const of the one picked by user? (if you let this possible...)

File: README.md [hunk 10]
```diff
@@ -117,4 +117,82 @@ require.False(t, testResult.Success)
...
+
+ensembleContainers, err := ensemble.RunContainers(ctx,
+    // Microcks container in ensemble
+    ensemble.WithMicrocksImage("quay.io/microcks/microcks-uber:latest"), // Optional
+    ensemble.WithMainArtifact("testdata/apipastries-openapi.yaml"),
+    ensemble.WithSecondaryArtifact("testdata/apipastries-postman-collection.json"),
+
+    // Postman container in ensemble
+    ensemble.WithPostman(true),
+    ensemble.WithPostmanImage("quay.io/microcks/microcks-postman-runtime:latest"),  // Optional
```
Comment (lbroudoux) [inline README.md:180]: Couldn't we deduce that `WithPostman()` is set to true if end-user is using `WithPostmanImage()`?

File: ensemble/ensemble.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+func WithPostman(enable bool) Option {
+	return func(e *MicrocksContainersEnsemble) error {
+		e.postmanEnabled = enable
+		return nil
+	}
+}
+
+// WithPostmanImage helps to use specific Postman image
+func WithPostmanImage(image string) Option {
+	return func(e *MicrocksContainersEnsemble) error {
```
Comment (lbroudoux) [inline ensemble/ensemble.go:172]: Why not enforcing `e.postmanEnabled = true` if this function is invoked?

**Review by lbroudoux — CHANGES_REQUESTED** [review]
It looks awesome! Thank you so much! I suggest one or two changes to have consistent approach for end-user but this is just to say something 😉

File: README.md (hunk 10 above)
Comment (JulienBreux) [reply README.md:180]: Yeah! Very good suggestion!

File: ensemble/ensemble.go (hunk 11 above)
Comment (JulienBreux) [reply ensemble/ensemble.go:172]: Yes, I'll do! Very good suggestion!

**Review by lbroudoux — APPROVED** [review]
Awesome! LGTM. 🎉

**Pull request merged**